File: bj/2805.cpp
#include<iostream>
#include<vector>
using namespace std;

int main(){
    long long n, m;
    cin >> n >> m;
    vector<long long> trees;
    long long max = 0;
    for (long long i = 0; i < n; i++){
        long long temp;
        cin >> temp;
        if (temp > max){
            max = temp;
        }
        trees.push_back(temp);
    }
    long long start = 0;
    long long end = max - 1;
    while(start <= end){
        long long mid = (start+end)/2;
        long long take = 0;
        for (long long j = 0; j < n; j++){
            if (trees[j] > mid){
                take += trees[j]-mid;
            }
        }
        if (take < m){
            end = mid - 1;
        }
        else{
            start = mid + 1;
        }
    }
    cout << end;
    return 0;
}

File: bj/2805.py
import sys
n, m = map(int, sys.stdin.readline().strip().split())
trees = list(sys.stdin.readline().strip().split())
trees = [int(tree) for tree in trees]
start = 0
end = max(trees)-1
while start <= end:
    mid = (start + end) // 2
    take = sum([trees[i]-mid if trees[i]-mid > 0 else 0 for i in range(n)])
    if take < m:
        end = mid - 1
    else:
        start = mid + 1
print(end)